### PR TITLE
Adding multiple labels feature

### PR DESCRIPTION
**Pull request by KharchenkoMaks**
- Each task now is able to have unlimited amount of labels.
- "label" command was renamed to "add_label" -> adds label to task by its id.
- Added "remove_label" command -> removes label from task by its id.

### CODE REVIEW

**Review by MichaelKupriianov — APPROVED** [review]

**Review by MichaelKupriianov — COMMENTED** [review]
.

**Review by MichaelKupriianov — CHANGES_REQUESTED** [review]
.

File: src/model/TaskManager.cpp [hunk 1]
```diff
@@ -129,16 +129,43 @@ std::optional<RelationalTask> TaskManager::GetTask(const TaskId& task_id) {
 }
 
 TaskActionResult TaskManager::AddTaskLabel(const TaskId& id, const std::string& label) {
+    if (label.empty())
+        return TaskActionResult::FAIL_LABEL_ALREADY_SET;
```
Comment (AntonApraksin) [inline src/model/TaskManager.cpp]: Irrelevant error

File: src/utilities/TaskComparators.cpp [hunk 2]
```diff
@@ -17,7 +17,8 @@ bool operator==(const Task& task1, const Task& task2) {
            task1.priority() == task2.priority() &&
            task1.due_date() == task2.due_date() &&
            task1.completed() == task2.completed() &&
-           task1.label() == task2.label();
+           task1.label_size() == task2.label_size() &&
+           std::equal(task1.label().begin(), task1.label().end(), task1.label().begin());
```
Comment (AntonApraksin) [inline src/utilities/TaskComparators.cpp]: task1 labels are compared to themselves

File: src/view/commands/CommandResult.cpp [hunk 3]
```diff
@@ -17,10 +18,11 @@ CommandResult CommandResult::Create(const TaskId& task_id) {
     return cmd_result;
 }
 
-CommandResult CommandResult::Create(const std::vector<RelationalTask>& tasks_to_show) {
+CommandResult CommandResult::Create(const std::vector<RelationalTask>& tasks_to_show, const bool show_tasks_relations) {
```
Comment (AntonApraksin) [inline src/view/commands/CommandResult.cpp]: Why pass const value?

File: src/view/user_interface/StateContext.cpp [hunk 4]
```diff
@@ -31,11 +31,11 @@ std::optional<TaskId> StateContext::GetTaskId() const {
 }
 
 void StateContext::AddTaskLabel(const std::string& task_label) {
```
Comment (AntonApraksin) [inline src/view/user_interface/StateContext.cpp]: SetTaskLabel is more suitalbe here

**Review by AntonApraksin — CHANGES_REQUESTED** [review]

File: src/utilities/TaskComparators.cpp (hunk 2 above)
Comment (KharchenkoMaks) [reply src/utilities/TaskComparators.cpp]: fixed

File: src/view/user_interface/StateContext.cpp (hunk 4 above)
Comment (KharchenkoMaks) [reply src/view/user_interface/StateContext.cpp]: fixed

File: src/model/TaskManager.cpp (hunk 1 above)
Comment (KharchenkoMaks) [reply src/model/TaskManager.cpp]: fixed to TaskActionResult::FAIL_INVALID_LABEL

File: src/view/commands/CommandResult.cpp (hunk 3 above)
Comment (KharchenkoMaks) [reply src/view/commands/CommandResult.cpp]: https://stackoverflow.com/questions/117293/use-of-const-for-function-parameters

**Review by antonovchtd — APPROVED** [review]

**Review by AntonApraksin — APPROVED** [review]

**Review by MichaelKupriianov — APPROVED** [review]

**Review by AntonApraksin — APPROVED** [review]

**Review by AntonApraksin — APPROVED** [review]

**Review by IlliaPlaksa — APPROVED** [review]
ControllerRequestResult can be renamed to ControllerResponse

**Review by antonovchtd — APPROVED** [review]

File: src/utilities/TaskBuilder.cpp [hunk 5]
```diff
@@ -29,8 +29,15 @@ TaskBuilder& TaskBuilder::SetCompletion(bool is_completed) {
...
 }
 
-TaskBuilder& TaskBuilder::SetLabel(const std::string &label) {
-    label_ = label;
+TaskBuilder& TaskBuilder::AddLabel(const std::string& label) {
+    labels_.push_back(label);
+    return *this;
+}
+
+TaskBuilder& TaskBuilder::RemoveLabel(const std::string& label) {
```
Comment (vladimir-moiseiev) [inline src/utilities/TaskBuilder.cpp]: A `Builder` pattern does not allow any "remove" logic. It's not a "Getter/Setter"
You create and fill it once to create an object. You can create multiple (same) objects, but you never change it to create another object.
If you "remove" label, you create another builder and don't set a label

@IlliaPlaksa @KharchenkoMaks @AntonApraksin @antonovchtd @MichaelKupriianov

File: src/view/user_interface/StateContext.cpp [hunk 6]
```diff
@@ -31,11 +31,11 @@ std::optional<TaskId> StateContext::GetTaskId() const {
...
-    task_.SetLabel(task_label);
+    task_label_ = task_label;
```
Comment (vladimir-moiseiev) [inline src/view/user_interface/StateContext.cpp:34]: is it Add or Set? Looks like Set

File: src/view/user_interface/UserInterface.cpp [hunk 7]
```diff
@@ -45,6 +45,9 @@ void UserInterface::PrintRequestResult(ControllerRequestResult action_result) {
         case ControllerRequestResult::FILE_DAMAGED:
             states_factory_->GetConsolePrinter()->WriteError("File is damaged.");
             break;
+        case ControllerRequestResult::FAIL_NO_SUCH_LABEL:
+            states_factory_->GetConsolePrinter()->WriteError("Fail to remove label, current task doesn't include such label.");
```
Comment (vladimir-moiseiev) [inline src/view/user_interface/UserInterface.cpp:49]: ConsolePrinter should be a direct dependency in your UI class. No need for factory.

**Review by vladimir-moiseiev — CHANGES_REQUESTED** [review]

File: src/utilities/TaskBuilder.cpp (hunk 5 above)
Comment (KharchenkoMaks) [reply src/utilities/TaskBuilder.cpp]: removed

File: src/view/user_interface/StateContext.cpp (hunk 6 above)
Comment (KharchenkoMaks) [reply src/view/user_interface/StateContext.cpp:34]: renamed to SetTaskLabel

File: src/view/user_interface/UserInterface.cpp (hunk 7 above)
Comment (KharchenkoMaks) [reply src/view/user_interface/UserInterface.cpp:49]: fixed in #6

**Review by vladimir-moiseiev — APPROVED** [review]

**Pull request merged**